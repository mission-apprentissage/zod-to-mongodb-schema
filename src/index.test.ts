import { JSONSchema7 } from "json-schema";
import { describe, expect, it } from "vitest";

import {
  jsonSchemaToMongoSchema,
  zodToMongoSchema,
  zObjectId,
} from "./index.ts";
import { z } from "zod";

describe("jsonSchemaToMongoSchema", () => {
  it("should convert complex schema", () => {
    const input: JSONSchema7 = {
      type: "object",
      properties: {
        _id: {
          $ref: "#/definitions/objectId",
        },
        type_document: {
          type: "string",
          description: "Le type de document (exemple: DECA, etc..)",
        },
        ext_fichier: {
          type: "string",
          description: "Le type de fichier extension",
          enum: ["xlsx", "xls", "csv"], // 10mb
        },
        nom_fichier: {
          type: "string",
          description: "Le nom de fichier",
        },
        chemin_fichier: {
          type: "string",
          description: "Chemin du fichier binaire",
        },
        taille_fichier: {
          type: "integer",
          description: "Taille du fichier en bytes",
        },
        hash_secret: {
          type: "string",
          description: "Hash fichier",
        },
        hash_fichier: {
          type: "string",
          description: "Checksum fichier",
        },
        import_progress: {
          type: "number",
          description: "Progress percentage (-1 not started)",
        },
        lines_count: {
          type: "integer",
          description: "Number of lines",
        },
        added_by: {
          type: "string",
          description: "Qui a ajouté le fichier",
        },
        updated_at: {
          type: "string",
          format: "date-time",
          description: "Date de mise à jour en base de données",
        },
        created_at: {
          type: "string",
          format: "date-time",
          description: "Date d'ajout en base de données",
        },
        _meta: {
          type: "object",
          additionalProperties: {},
        },
      },
      required: [
        "_id",
        "type_document",
        "ext_fichier",
        "nom_fichier",
        "chemin_fichier",
        "taille_fichier",
        "hash_secret",
        "hash_fichier",
        "added_by",
        "created_at",
      ],
      additionalProperties: false,
    };

    expect(jsonSchemaToMongoSchema(input, input)).toEqual({
      bsonType: "object",
      properties: {
        _id: {
          bsonType: "objectId",
        },
        type_document: {
          bsonType: "string",
          description: "Le type de document (exemple: DECA, etc..)",
        },
        ext_fichier: {
          bsonType: "string",
          description: "Le type de fichier extension",
          enum: ["xlsx", "xls", "csv"], // 10mb
        },
        nom_fichier: {
          bsonType: "string",
          description: "Le nom de fichier",
        },
        chemin_fichier: {
          bsonType: "string",
          description: "Chemin du fichier binaire",
        },
        taille_fichier: {
          bsonType: "int",
          description: "Taille du fichier en bytes",
        },
        hash_secret: {
          bsonType: "string",
          description: "Hash fichier",
        },
        hash_fichier: {
          bsonType: "string",
          description: "Checksum fichier",
        },
        import_progress: {
          bsonType: "number",
          description: "Progress percentage (-1 not started)",
        },
        lines_count: {
          bsonType: "int",
          description: "Number of lines",
        },
        added_by: {
          bsonType: "string",
          description: "Qui a ajouté le fichier",
        },
        updated_at: {
          bsonType: "date",
          description: "Date de mise à jour en base de données",
        },
        created_at: {
          bsonType: "date",
          description: "Date d'ajout en base de données",
        },
        _meta: {
          bsonType: "object",
          additionalProperties: true,
        },
      },
      required: [
        "_id",
        "type_document",
        "ext_fichier",
        "nom_fichier",
        "chemin_fichier",
        "taille_fichier",
        "hash_secret",
        "hash_fichier",
        "added_by",
        "created_at",
      ],
      additionalProperties: false,
    });
  });
});

describe("zodToMongoSchema", () => {
  it("should convert zod object properly", () => {
    expect(
      zodToMongoSchema(
        z
          .object({
            _id: zObjectId,
            string: z.string(),
            number: z.number(),
            enum: z.enum(["a", "b"]),
            optionalString: z.string().optional(),
            nullishNumber: z.number().nullish(),
          })
          .strict(),
      ),
    ).toMatchSnapshot();
  });

  it("should resolves zod ref properly", () => {
    const zEmail = z.string().email();

    expect(
      zodToMongoSchema(
        z
          .object({
            _id: zObjectId,
            a: zEmail,
            b: zEmail,
          })
          .strict(),
      ),
    ).toMatchSnapshot();
  });
});
