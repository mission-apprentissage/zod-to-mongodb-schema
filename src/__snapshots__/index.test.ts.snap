// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zodToMongoSchema > should convert zod object properly 1`] = `
{
  "additionalProperties": false,
  "bsonType": "object",
  "properties": {
    "_id": {
      "bsonType": "objectId",
    },
    "enum": {
      "bsonType": "string",
      "enum": [
        "a",
        "b",
      ],
    },
    "nullishNumber": {
      "bsonType": [
        "number",
        "null",
      ],
    },
    "number": {
      "bsonType": "number",
    },
    "optionalString": {
      "bsonType": "string",
    },
    "string": {
      "bsonType": "string",
    },
  },
  "required": [
    "_id",
    "string",
    "number",
    "enum",
  ],
}
`;

exports[`zodToMongoSchema > should resolves zod ref properly 1`] = `
{
  "additionalProperties": false,
  "bsonType": "object",
  "properties": {
    "_id": {
      "bsonType": "objectId",
    },
    "a": {
      "bsonType": "string",
    },
    "b": {
      "bsonType": "string",
    },
  },
  "required": [
    "_id",
    "a",
    "b",
  ],
}
`;
